pkg/runtime/sequencer/cache.go

```diff
@@ -2,15 +2,198 @@ package sequencer
+// When InUseBuffer is used up, swap them.
+type DoubleBuffer struct {
+	Key          string
+	Size         int
+	InUseBuffer  *Buffer
+	BackUpBuffer *Buffer
+	lock         sync.Mutex
+	Store        sequencer.Store
+}
+
+type Buffer struct {
+	from int64
+	to   int64
+}
+
+func NewDoubleBuffer(key string, store sequencer.Store) *DoubleBuffer {
+
+	d := &DoubleBuffer{
+		Key:   key,
+		Size:  defaultSize,
+		Store: store,
+	}
+
+	return d
+}
+
+//init double buffer
+func (d *DoubleBuffer) init() error {
+
+	buffer, err := d.getNewBuffer()
+	if err != nil {
+		return err
+	}
+
+	d.InUseBuffer = buffer
+
+	return nil
+}
+
+//getId next id
+func (d *DoubleBuffer) getId() (int64, error) {
+
+	if d.InUseBuffer == nil {
+		return 0, errors.New("[DoubleBuffer] Get error: InUseBuffer nil ")
+	}
+
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	//check swap
+	if d.InUseBuffer.from > d.InUseBuffer.to {
+		err := d.swap()
+		if err != nil {
+			return 0, err
+		}
+	}
+	next := d.InUseBuffer.from
+	d.InUseBuffer.from++
+
+	//when InUseBuffer id more than half used, initialize BackUpBuffer.
+	//equal make sure only one thread enter
+	if d.InUseBuffer.to-d.InUseBuffer.from == defaultLimit {
+		utils.GoWithRecover(func() {
+
+			//lock to avoid data race
+			d.lock.Lock()
+			defer d.lock.Unlock()
+			//one case: here and swap are performed simultaneously,
+			//swap fast,no check will cover old BackUpBuffer
+			if d.BackUpBuffer != nil {
+				return
+			}
+			buffer, err := d.getNewBuffer()
+			if err != nil {
+				log.DefaultLogger.Errorf("[DoubleBuffer] [getNewBuffer] error: %v", err)
+				return
+			}
+
+			d.BackUpBuffer = buffer
+
+		}, nil)
+	}
+
+	return next, nil
+}
+
+//swap InUseBuffer and BackUpBuffer, must be locked
+func (d *DoubleBuffer) swap() error {
+
+	//check again
+	if d.BackUpBuffer == nil {
+		buffer, err := d.getNewBuffer()
+		if err != nil {
+			return err
+		}
+		d.BackUpBuffer = buffer
+	}
+
+	d.InUseBuffer = d.BackUpBuffer
+	d.BackUpBuffer = nil
+	return nil
+}
+
+//getNewBuffer return a new segment
+func (d *DoubleBuffer) getNewBuffer() (*Buffer, error) {
+	support, result, err := d.Store.GetSegment(&sequencer.GetSegmentRequest{
+		Key:  d.Key,
+		Size: d.Size,
+	})
+	if err != nil {
+		return nil, err
+	}
+	if !support {
+		return nil, errors.New("[DoubleBuffer] unSupport Segment id")
+	}
+	return &Buffer{
+		from: result.From,
+		to:   result.To,
+	}, nil
+}
+
+var BufferCatch = map[string]*DoubleBuffer{}
+
+//common lock is enough ？
+var rwLock sync.RWMutex
+
 func GetNextIdFromCache(ctx context.Context, store sequencer.Store, req *sequencer.GetNextIdRequest) (bool, int64, error) {
-	// TODO
-	//	1. check cache
-	//	2. load cache with lock
-	//	2.1. return if not support
-	//	2.2. add segment into cache
-	//	2.3. return result
-	return false, 0, nil
+
+	//check support
+	support, _, _ := store.GetSegment(&sequencer.GetSegmentRequest{
```

Comment:
access store every time is better to be prevented

Reply:
@MoonShining  `size=0` will only return true/false  , will not request components, such as redis

Reply:
so the behaviour rely on the internal implementation in specific store. other stores may not care about size=0.

is it better to expose some ability check interface on store? eg: `store.Support("SegmentGet")`

Reply:
@MoonShining It's a long story...The first version of the component interface definition has a support() method, but after discussion I removed it ,see https://github.com/mosn/layotto/issues/158#issuecomment-890322241
So actually this invocation here just means "check whether this component support it"
If u guys also think it is weird,I will submit a PR to add the 'support' method back. But I think it's ok here in this PR